UBSAN: shift exponent 513 is too large for 32-bit type 'int'

UB sanitizer catches excessive shifting in `tfw_gfsm_move()`:

```
[   51.679524] UBSAN: Undefined behaviour in /root/tempesta/tempesta_fw/gfsm.c:226:25
[   51.683145] shift exponent 513 is too large for 32-bit type 'int'
[   51.685696] CPU: 2 PID: 0 Comm: swapper/2 Tainted: G           O    4.14.32-tfw-asan #0.0.0
[   51.689817] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS 1.12.0-1 04/01/2014
[   51.695066] Call Trace:
[   51.697234]  <IRQ>
[   51.699067]  dump_stack+0x90/0xd0
[   51.700904]  ? _atomic_dec_and_lock+0x1b0/0x1b0
[   51.703034]  ubsan_epilogue+0x9/0x7c
[   51.704900]  __ubsan_handle_shift_out_of_bounds+0x24c/0x291
[   51.707311]  ? __ubsan_handle_load_invalid_value+0x163/0x163
[   51.709794]  ? ttls_recv+0x134c/0x1d40 [tempesta_tls]
[   51.712063]  ? tfw_gfsm_move+0x916/0x9f0 [tempesta_fw]
[   51.714522]  tfw_gfsm_move+0x916/0x9f0 [tempesta_fw]
[   51.716907]  ? ttls_decrypt+0x1830/0x1830 [tempesta_tls]
[   51.719264]  ? __gfsm_fsm_exec+0x3d0/0x3d0 [tempesta_fw]
[   51.731211]  ? ip_output+0xc0/0x320
[   51.733664]  tfw_tls_msg_process+0x49d/0xde0 [tempesta_fw]
[   51.736408]  ? tfw_tls_conn_dtor+0x630/0x630 [tempesta_fw]
[   51.738810]  ? ipt_do_table+0x938/0x14f0
[   51.740939]  ? ip_finish_output2+0x5e1/0x11c0
[   51.743153]  ? ip_copy_metadata+0x720/0x720
[   51.745202]  ? tfw_tls_conn_dtor+0x630/0x630 [tempesta_fw]
[   51.747581]  __gfsm_fsm_exec+0x1ea/0x3d0 [tempesta_fw]
[   51.749885]  ? tfw_gfsm_register_hook+0x8d0/0x8d0 [tempesta_fw]
[   51.752345]  ? ip_output+0x1bc/0x320
[   51.754221]  ? ip_output+0x1bc/0x320
[   51.756092]  tfw_connection_recv+0x78/0x90 [tempesta_fw]
[   51.758412]  ? tfw_connection_send+0xe0/0xe0 [tempesta_fw]
[   51.761894]  ss_tcp_process_data+0x721/0x1170 [tempesta_fw]
[   51.764990]  ? ss_getpeername+0x790/0x790 [tempesta_fw]
[   51.767708]  ? __tcp_select_window+0x800/0x800
[   51.770041]  ? pg_skb_alloc+0xf70/0xf70
[   51.772252]  ss_tcp_data_ready+0xd9/0x280 [tempesta_fw]
[   51.774829]  tcp_rcv_established+0xc81/0x1c90
[   51.777045]  ? save_stack+0x33/0xa0
[   51.779019]  ? tcp_data_queue+0x3e30/0x3e30
[   51.781184]  ? sk_filter_trim_cap+0x65/0x1030
[   51.783399]  ? tcp_v4_reqsk_send_ack+0x440/0x440
[   51.785690]  ? do_idle+0x1ca/0x290
[   51.787631]  tcp_v4_do_rcv+0x3cb/0x890
[   51.789650]  ? tcp_v4_fill_cb+0x28/0x500
[   51.792949]  tcp_v4_rcv+0x1ddb/0x2fd0
[   51.795307]  ? tcp_v4_early_demux+0x830/0x830
[   51.797418]  ? do_add_counters+0x720/0x720
[   51.799434]  ? raw_rcv+0x3a0/0x3a0
[   51.801471]  ? __inet_lookup_established+0x50/0x6d0
[   51.803821]  ip_local_deliver_finish+0x207/0x810
[   51.805992]  ? do_add_counters+0x720/0x720
[   51.807998]  ip_local_deliver+0x19a/0x2f0
[   51.810101]  ? ip_call_ra_chain+0x850/0x850
[   51.812278]  ? tcp_v4_early_demux+0x642/0x830
[   51.814524]  ? ip_rcv_finish+0x11d0/0x11d0
[   51.816672]  ? ip_rcv_finish+0xc0a/0x11d0
[   51.818808]  ip_rcv+0x81b/0x10e0
[   51.821647]  ? packet_rcv+0x187/0x2120
[   51.824146]  ? ip_local_deliver+0x2f0/0x2f0
[   51.826230]  ? inet_del_offload+0x30/0x30
[   51.828259]  ? ip_local_deliver+0x2f0/0x2f0
[   51.830320]  __netif_receive_skb_core+0x11ba/0x3c20
[   51.832682]  ? nf_ingress+0x760/0x760
[   51.834729]  ? virtqueue_detach_unused_buf+0x220/0x220 [virtio_ring]
[   51.837352]  ? virtqueue_napi_complete+0x34/0x60 [virtio_net]
[   51.839774]  ? virtnet_poll+0x628/0xb23 [virtio_net]
[   51.842008]  ? process_backlog+0x229/0x8d0
[   51.844009]  process_backlog+0x229/0x8d0
[   51.845977]  ? __netif_receive_skb+0x220/0x220
[   51.848065]  ? vring_interrupt+0x139/0x1d0 [virtio_ring]
[   51.851221]  net_rx_action+0x7db/0xf30
[   51.853142]  ? napi_complete_done+0x3e0/0x3e0
[   51.855234]  ? try_to_wake_up+0xc3/0x1690
[   51.857222]  ? handle_irq_event+0x144/0x270
[   51.859426]  ? timer_migration_handler+0x70/0x70
[   51.861700]  ? migrate_swap_stop+0x860/0x860
[   51.863892]  ? handle_edge_irq+0x23e/0xd80
[   51.866074]  __do_softirq+0x201/0x799
[   51.867992]  irq_exit+0x110/0x160
[   51.869822]  do_IRQ+0x7a/0x1c0
[   51.871562]  common_interrupt+0x7a/0x7a
[   51.873501]  </IRQ>
[   51.874996] RIP: 0010:default_idle+0x15/0x2c0
[   51.877057] RSP: 0018:ffff880139827ee8 EFLAGS: 00000246 ORIG_RAX: ffffffffffffffdd
[   51.881564] RAX: ffffffffb18c1070 RBX: ffff880139802c40 RCX: 0000000000000000
[   51.884359] RDX: 1ffff10027300588 RSI: 0000000000000000 RDI: 0000000000000000
[   51.887149] RBP: ffffed0027300588 R08: 1ffffffff64c1200 R09: dffffc0000000000
[   51.890044] R10: 0000000000004d62 R11: 0000000000000001 R12: ffff880139802c40
[   51.892888] R13: 0000000000000002 R14: 0000000000000000 R15: 0000000000000000
[   51.895918]  ? __sched_text_end+0x6/0x6
[   51.897915]  do_idle+0x1ca/0x290
[   51.899700]  cpu_startup_entry+0x14/0x20
[   51.901696]  secondary_startup_64+0xa5/0xb0
[   51.903717] ================================================================================
```

Here is where UBSAN is pointing at:
```c
int
tfw_gfsm_move(TfwGState *st, unsigned short state, TfwFsmData *data)
{
        int r = TFW_PASS, p, fsm;
        unsigned int *hooks = fsm_hooks_bm[FSM(st)];
        unsigned long mask = 1 << state;  /* ← this */
        unsigned char curr_st = st->curr;

        TFW_DBG3("GFSM move %#x -> %#x: skb=%pK off=%u trail=%u"
                 " req=%pK resp=%pK\n", FSM_STATE(st), state,
                 data->skb, data->off, data->trail, data->req, data->resp);
```

**Upd. (22 Apr 2019)**

Also, the same issue for `tfw_gfsm_move()` called from `frang_resp_handler()`.

According to CLang the issue in the TLS FSM states definitions: 
```
In file included from tempesta/tls/tls_srv.c:28:
tempesta/tls/tls_internal.h:416:21: warning: signed shift result (0x6000000000000) requires 52 bits to
      represent, but 'int' only has 32 bits [-Wshift-overflow]
        TTLS_CC_HS_ALLOC        = __TTLS_FSM_SUBST(CLIENT_CERTIFICATE, 0),
                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/mnt/Storage/Projects/tempesta/tempesta/tls/tls_internal.h:340:36: note: expanded from macro '__TTLS_FSM_SUBST'
#define __TTLS_FSM_SUBST(st, sst)       (__TTLS_FSM_ST(TTLS_##st) | sst)
                                         ^~~~~~~~~~~~~~~~~~~~~~~~
/mnt/Storage/Projects/tempesta/tempesta/tls/tls_internal.h:338:34: note: expanded from macro '__TTLS_FSM_ST'
#define __TTLS_FSM_ST(st)               ((st) << __TTLS_FSM_ST_SHIFT)
                                         ~~~~ ^  ~~~~~~~~~~~~~~~~~~~
```

--- a/tempesta_fw/gfsm.c
+++ b/tempesta_fw/gfsm.c
@@ -84,10 +84,6 @@
 				 & TFW_GFSM_PRIO_MASK)
 #define PRIO(s)			__GFSM_PRIO(FSM_STATE(s))
 #define GFSM_HOOK_N		(TFW_GFSM_PRIO_N * TFW_GFSM_STATE_N)
-#define SET_STATE(s, x)						\
-do {								\
-	FSM_STATE(s) = (FSM_STATE(s) & ~TFW_GFSM_STATE_MASK) | (x); \
-} while (0)
 
 #define __BAD_STATE_BYTE	0xff
 #define BAD_STATE		((__BAD_STATE_BYTE << 8) | __BAD_STATE_BYTE)
@@ -223,14 +219,15 @@ tfw_gfsm_move(TfwGState *st, unsigned short state, TfwFsmData *data)
 {
 	int r = TFW_PASS, p, fsm;
 	unsigned int *hooks = fsm_hooks_bm[FSM(st)];
-	unsigned long mask = 1 << state;
+	unsigned long mask = 1 << (state & TFW_GFSM_STATE_MASK);
 	unsigned char curr_st = st->curr;
 
 	T_DBG3("GFSM move %#x -> %#x: skb=%pK req=%pK resp=%pK\n",
 	       FSM_STATE(st), state, data->skb, data->req, data->resp);
 
 	/* Remember current FSM context. */
-	SET_STATE(st, state);
+	FSM_STATE(st) = (FSM_STATE(st) & ~TFW_GFSM_STATE_MASK) |
+			(state & TFW_GFSM_STATE_MASK);
 
 	/* Start from highest priority. */
 	for (p = TFW_GFSM_HOOK_PRIORITY_HIGH;

--- a/tempesta_fw/vhost.c
+++ b/tempesta_fw/vhost.c
@@ -1522,9 +1522,10 @@ tfw_vhost_create(const char *name)
 {
 	TfwPool *pool;
 	TfwVhost *vhost;
-	int name_sz = strlen(name) + 1;
+	int name_strlen = strlen(name);
+	int name_mem_sz = ALIGN(name_strlen + 1, sizeof(void *));
 	int size = sizeof(TfwVhost)
-		+ name_sz
+		+ name_mem_sz
 		+ sizeof(TfwLocation) * (TFW_LOCATION_ARRAY_SZ + 1)
 		+ sizeof(FrangGlobCfg)
 		+ tfw_tls_vhost_priv_data_sz();
@@ -1539,8 +1540,8 @@ tfw_vhost_create(const char *name)
 	}
 	INIT_HLIST_NODE(&vhost->hlist);
 	vhost->name.data = (char *)(vhost + 1);
-	vhost->name.len = name_sz - 1;
-	vhost->loc_dflt = (TfwLocation *)(vhost->name.data + name_sz);
+	vhost->name.len = name_strlen;
+	vhost->loc_dflt = (TfwLocation *)(vhost->name.data + name_mem_sz);
 	vhost->loc = (TfwLocation *)(vhost->loc_dflt + 1);
 	vhost->frang_gconf = (FrangGlobCfg *)(vhost->loc + TFW_LOCATION_ARRAY_SZ);
 	vhost->tls_cfg.priv = (vhost->frang_gconf + 1);
@@ -1549,7 +1550,7 @@ tfw_vhost_create(const char *name)
 	BUG_ON((char *)vhost->tls_cfg.priv + tfw_tls_vhost_priv_data_sz() !=
 	       (char *)vhost + size);
 
-	memcpy(vhost->name.data, name, name_sz);
+	memcpy(vhost->name.data, name, name_strlen + 1);
 	vhost->hdrs_pool = pool;
 	atomic64_set(&vhost->refcnt, 1);
 

--- a/tempesta_fw/vhost.c
+++ b/tempesta_fw/vhost.c
@@ -1526,6 +1526,7 @@ tfw_vhost_create(const char *name)
 	int size = sizeof(TfwVhost)
 		+ name_sz
 		+ sizeof(TfwLocation) * (TFW_LOCATION_ARRAY_SZ + 1)
+		+ sizeof(FrangGlobCfg)
 		+ tfw_tls_vhost_priv_data_sz();
 
 	if (!(pool = __tfw_pool_new(0)))
@@ -1543,6 +1544,11 @@ tfw_vhost_create(const char *name)
 	vhost->loc = (TfwLocation *)(vhost->loc_dflt + 1);
 	vhost->frang_gconf = (FrangGlobCfg *)(vhost->loc + TFW_LOCATION_ARRAY_SZ);
 	vhost->tls_cfg.priv = (vhost->frang_gconf + 1);
+
+	/* Must be sure all data fits, to prevent silent data corruption. */
+	BUG_ON((char *)vhost->tls_cfg.priv + tfw_tls_vhost_priv_data_sz() !=
+	       (char *)vhost + size);
+
 	memcpy(vhost->name.data, name, name_sz);
 	vhost->hdrs_pool = pool;
 	atomic64_set(&vhost->refcnt, 1);

